### PR TITLE
Fix swapping evaluation order in assignments with CSE defs

### LINKED ISSUE #57121: ARM64: Another null ref, this time maybe without VN
```csharp
// Generated by Fuzzlyn v1.2 on 2021-08-09 10:31:59
// Run on .NET 6.0.0-dev on Arm64 Linux
// Seed: 16685661424470671852
// Reduced from 36.1 KiB to 0.4 KiB in 00:01:03
// Debug: Runs successfully
// Release: Throws 'System.NullReferenceException'
public class Program
{
    static short s_2;
    static bool s_7;
    public static void Main()
    {
        var vr4 = new ulong[][][][]{new ulong[][][]{new ulong[][]{new ulong[]{0}}}};
        int vr9 = 866278096;
        s_2 |= (short)(0 % (vr9 | 1));
        if (s_7)
        {
            vr4[0] = vr4[0];
        }

        System.Console.WriteLine(vr9);
    }
}
```

cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by jakobbotsch**
In some cases we were swapping the evaluation order of operands in
assignments when there were CSE defs in the subtrees. Change the logic
to properly forego the swapping in this situation.

Fix #57121

cc @dotnet/jit-contrib

### CODE REVIEW

**Comment by jakobbotsch** [comment]
I'll post diffs later.

**Review by sandreenko — APPROVED** [review]
LGTM

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -4205,7 +4205,7 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
     DONE_OP1_AFTER_COST:
 
         bool bReverseInAssignment = false;
-        if (oper == GT_ASG)
+        if (oper == GT_ASG && (!optValnumCSE_phase || optCSE_canSwap(op1, op2)))
```
Comment (briansull) [inline src/coreclr/jit/gentree.cpp:4208]: I see that you also use optCSE_canSwap,  that allows us to reorder more cases and should result in the minimal diffs.
Probably only the broken cases would have diffs.

**Review by briansull — APPROVED** [review]

**Comment by jakobbotsch** [comment]
No SPMI diffs and infra failures look unrelated.

**Pull request merged**